leftwm-core/src/handlers/command_handler/scratchpad_handler.rs

```diff
@@ -0,0 +1,1242 @@
+
+use serde::{Deserialize, Serialize};
+
+use crate::{
+    child_process::exec_shell,
+    handlers::window_handler::scratchpad_xyhw,
+    models::{TagId, WindowHandle},
+    Command, Config, DisplayAction, DisplayServer, Manager, Window,
+};
+
+/// Describes the options for the release scratchpad command
+#[derive(Serialize, Deserialize, Clone, PartialEq, Eq, Debug)]
+pub enum ReleaseScratchPadOption {
+    /// Release a window from a scratchpad given a window handle
+    Handle(WindowHandle),
+    /// Release a window from a scratchpad given a scratchpad name, the most upper window in the
+    /// scratchpad queue will be released
+    ScrathpadName(String),
+    /// Release the currently focused window from its scratchpad
+    None,
+}
+
+/// Hide scratchpad window:
+/// Expects that the window handle is a valid handle to a visible scratchpad window
+fn hide_scratchpad<C: Config, SERVER: DisplayServer>(
+    manager: &mut Manager<C, SERVER>,
+    scratchpad_window: &WindowHandle,
+) -> Result<(), &'static str> {
+    log::trace!("Hide scratchpad window {:?}", scratchpad_window);
+    let nsp_tag = manager
+        .state
+        .tags
+        .get_hidden_by_label("NSP")
+        .ok_or("Could not find NSP tag")?;
+    let window = manager
+        .state
+        .windows
+        .iter_mut()
+        .find(|w| w.handle == *scratchpad_window)
+        .ok_or("Could not find window from scratchpad_window")?;
+
+    window.untag();
+    // Hide the scratchpad.
+    window.tag(&nsp_tag.id);
+    window.set_visible(false);
+
+    // send tag changement to X
+    let act = DisplayAction::SetWindowTag(*scratchpad_window, window.tag);
+    manager.state.actions.push_back(act);
+    manager.state.sort_windows();
+
+    // Will ignore current window handler because we just set it invisible
+    let last_focused_still_visible = manager
+        .state
+        .focus_manager
+        .window_history
+        .iter()
+        .find(|handle| {
+            if let Some(visibilty) = manager
+                .state
+                .windows
+                .iter()
+                .find(|window| Some(window.handle) == **handle)
+                .map(Window::visible)
+            {
+                visibilty
+            } else {
+                false
+            }
+        })
+        .copied();
+
+    // Make sure when changing focus the lastly focused window is focused
+    let handle = if let Some(prev) = last_focused_still_visible {
+        prev
+    } else if let Some(ws) = manager
+        .state
+        .focus_manager
+        .workspace(&manager.state.workspaces)
+    {
+        manager
+            .state
+            .windows
+            .iter()
+            .find(|w| ws.is_managed(w))
+            .map(|w| w.handle)
+    } else {
+        None
+    };
+    if let Some(handle) = handle {
+        manager.state.handle_window_focus(&handle);
+    }
+
+    Ok(())
+}
+
+/// Inverse of `hide_scratchpad` and makes the scratchpad visible
```

Comment:
This comment should relate to itself rather than another function. Just say it shows a scratchpad